libusb/descriptor.c

```diff
@@ -1517,3 +1517,90 @@ void API_EXPORTED libusb_free_interface_association_descriptors(
 		free((void*)iad_array->iad);
 	free(iad_array);
 }
+
+/** \ingroup libusb_desc
+ * Retrieve a device string without needing to open the device.
+ *
+ * \param dev the target device
+ * \param string_type the string type to retrieve
+ * \param data the data buffer for the UTF-8 encoded string.
+ * \param length the size of the data buffer in bytes.  
+ *      USB string descriptors cannot be longer than 
+ *      LIBUSB_DEVICE_STRING_UTF8_BYTES_MAX.
+ * \returns an error code or
+ *      the actual string length in bytes including the null terminator.
+ * \see libusb_get_string_descriptor()
+ * \see libusb_get_string_descriptor_ascii()
+ * 
+ * This function works when the device is still closed since it relies
+ * on the operating system to provide the string.  The operating system
+ * normally reads and caches the common string descriptors during
+ * USB enumeration.
+ *
+ * Since the USB string descriptor could be processed by the OS,
+ * this function returns a UTF-8 encoded string.  The string will
+ * be returned untranslated or in the default OS language
+ * when supported by the OS and USB device.
+ * 
+ * One way to call this function is using a buffer on the stack:
+ * 
+ *     char buffer[LIBUSB_DEVICE_STRING_UTF8_BYTES_MAX];
+ *     int ret = libusb_get_device_string(dev, LIBUSB_DEVICE_STRING_SERIAL_NUMBER, buffer, sizeof(buffer));
+ *     if (ret < 0) {
+ *         // handle error
+ *     }
+ * 
+ * This function is commonly used to get the serial number to allow
+ * for device selection before opening the selected device.
+ */
+int API_EXPORTED libusb_get_device_string(libusb_device* dev,
+	enum libusb_device_string_type string_type, char* data, int length) {
```

Comment:
Wouldn't `size_t` be better for length?

Reply:
I don't know, and signed vs unsigned length comes up repeatedly with C. We return int as string length or negative error. For consistency with this function, shouldn't the length argument be the same type?  

However, even more important is consistency with the API.  `libusb_get_string_descriptor` uses `int` length, so we should probably match this.  However, the use of signed vs unsigned length is inconsistent in libusb.  I looked through the libusb API:

Length signed:

* libusb_transfer
* libusb_fill_*_transfer (except control)
* libusb_*_transfer (except control)
* libusb_get_descriptor  <--- !!!
* libusb_get_string_descriptor & libusb_get_string_descriptor_ascii <--- !!!

Length unsigned:

* libusb_get_port_path (uint8_t)
* libusb_dev_mem_alloc & libusb_dev_mem_free
* libusb_fill_control_setup (uint16_t)
* libusb_set_iso_packet_lengths & all ISO functions
* libusb_control_transfer (uint16_t)

I recommend keeping this `int` to match `libusb_get_descriptor` and `libusb_get_string_descriptor`.  Do you agree?

Reply:
Sure. Maybe just add a check at the beginning of the function and return 'parameter error' if length is negative.

Reply:
Will add a (length < 0) check to `libusb_get_device_string`.